### PR TITLE
BUG: Index.intersection casting to object instead of numeric

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -2826,8 +2827,9 @@ def intersection(self, other, sort=False):
             return self._get_reconciled_name_object(other)
 
         if not is_dtype_equal(self.dtype, other.dtype):
-            this = self.astype("O")
-            other = other.astype("O")
+            dtype = find_common_type([self.dtype, other.dtype])
+            this = self.astype(dtype)
```
Comment (jreback) [inline pandas/core/indexes/base.py:2831]: prob nbd but can do copy=False

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm pls add a whatsnew note (1.2 fine)

**Comment by jbrockmendel** [comment]
already has whatsnew.  mind if i add the copy=False in the next pass?  lots in the pipeline

**Comment by jreback** [comment]
> already has whatsnew. mind if i add the copy=False in the next pass? lots in the pipeline

sure

**Review by jreback — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
I don't know if it was this PR (from this range https://github.com/pandas-dev/pandas/compare/65319af6...f04ec866), but the `intersection` method got a bit slowdown in case of numeric indexes: https://pandas.pydata.org/speed/pandas/#index_object.SetOperations.time_operation?p-dtype='int'&p-method='intersection'

**Comment by jorisvandenbossche** [comment]
Ah, might also be https://github.com/pandas-dev/pandas/pull/36927

**Pull request merged**